File: code/android-core-library/src/main/java/com/adobe/marketing/mobile/FileUtil.java
/*
  Copyright 2022 Adobe. All rights reserved.
  This file is licensed to you under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License. You may obtain a copy
  of the License at http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software distributed under
  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  OF ANY KIND, either express or implied. See the License for the specific language
  governing permissions and limitations under the License.
 */

package com.adobe.marketing.mobile;

import java.io.*;

final class FileUtil {

	private final static String LOG_TAG = "FileUtil";

	private FileUtil() {}

	/**
	 * Reads the JSONString from the provided file. Returns null if there is no file or it does not have read permissions.
	 *
	 * @param file {@link File} from which the contents are to be read
	 * @return The contents of the file in JSONString format. Returns null if file does not exist, when encounters IOException
	 * 			or if the file do not have read permission
	 */
	static String readStringFromFile(final File file) {

		try {
			if (file == null || !file.exists() || !file.canRead() || !file.isFile()) {
				Log.warning(LOG_TAG, "Write to file - File does not exist or don't have read permission (%s)", file);
				return null;
			}
		} catch (SecurityException e) {
			Log.debug(LOG_TAG, "Failed to read file (%s)", e);
			return null;
		}

		BufferedReader bufferedReader = null;
		InputStream inStream = null;

		try {
			inStream = new FileInputStream(file);
			final Reader reader = new InputStreamReader(inStream, "UTF-8");
			bufferedReader = new BufferedReader(reader);
			final StringBuilder builder = new StringBuilder();
			String line = bufferedReader.readLine();

			while (line != null) {
				builder.append(line);
				line = bufferedReader.readLine();
			}

			return builder.toString();
		} catch (IOException e) {
			Log.debug(LOG_TAG, "Failed to close file (%s)", e);
			return null;
		} finally {
			try {
				if (inStream != null) {
					inStream.close();
				}

				if (bufferedReader != null) {
					bufferedReader.close();
				}
			} catch (IOException e) {
				Log.debug(LOG_TAG, "Failed to close file (%s)", e);
			}
		}
	}

	/**
	 * Verifies if the {@code File} object represents a directory and the directory is writable.
	 *
	 * @param directory {@link File} the directory to check for
	 * @return {code boolean} representing the directory validation result
	 */
	static boolean isValidDirectory(final File directory) {
		return directory != null && directory.isDirectory() && directory.canWrite();
	}

}

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/services/utility/FileUtil.kt
/*
  Copyright 2022 Adobe. All rights reserved.
  This file is licensed to you under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License. You may obtain a copy
  of the License at http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software distributed under
  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  OF ANY KIND, either express or implied. See the License for the specific language
  governing permissions and limitations under the License.
 */
package com.adobe.marketing.mobile.services.utility

import android.content.Context
import com.adobe.marketing.mobile.LoggingMode
import com.adobe.marketing.mobile.MobileCore
import java.io.File

internal object FileUtil {

    private const val LOG_TAG = "FileUtil"

    /**
     * Removes the relative part of the file name(if exists).
     *
     * for ex: File name `/mydatabase/../../database1` will be converted to `mydatabase_database1`
     *
     * @param filePath the file name
     * @return file name without relative path
     */
    @JvmStatic
    fun removeRelativePath(filePath: String): String {
        return if (filePath.isEmpty()) {
            filePath
        } else {
            var result = filePath.replace("\\.[/\\\\]".toRegex(), "\\.")
            result = result.replace("[/\\\\](\\.{2,})".toRegex(), "_")
            result = result.replace("/".toRegex(), "")
            result
        }
    }

    /**
     * Copies the contents from `src` to `dest`.
     *
     * @param src [File] from which the contents are read
     * @param dest [File] to which contents are written to
     * @throws Exception if `src` or `dest` is not present or it does not have read permissions
     */
    @JvmStatic
    @Throws(Exception::class)
    fun copyFile(src: File, dest: File) {
        src.copyTo(dest, true)
    }

    @JvmStatic
    fun migrateSharedPreference(context: Context, dataStoreName: String, prefix: String) {
        val existingSharedPreferenceFile = File(getSharedPreferenceFilePath(context, dataStoreName))
        if (existingSharedPreferenceFile.exists()) {
            val newSharedPreferenceFile =
                File(getSharedPreferenceFilePath(context, prefix + dataStoreName))
            try {
                copyFile(existingSharedPreferenceFile, newSharedPreferenceFile)
                existingSharedPreferenceFile.delete()
            } catch (e: java.lang.Exception) {
                MobileCore.log(LoggingMode.DEBUG,
                    LOG_TAG,
                    "Unable to migrate SharedPreference file $e"
                )
            }
        }
    }

    private fun getSharedPreferenceFilePath(context: Context, dataStoreName: String): String {
        return "/data/data" + context.packageName + "/shared_prefs/" + dataStoreName + ".xml"
    }
}
